---
title: "Estimate model from Mostad et al."
author: "Malte BÃ¸dkergaard Nielsen"
date: "2025-03-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggnewscale) # Enables multiple colour- and fill-scales
library(latex2exp)
library(patchwork)
library(EnvStats)
source("00-functions-and-global-definitions.R")

# The palette with grey:
cbp1 <- c("#999999", "#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

```{r Reading data}
dd_GT <- readRDS(file.path("..", "data", "00_dd_GT.rds"))
dd_GT_2nd <- readRDS(file.path("..", "data", "00_dd_GT_2nd.rds"))

dd_GT$ID <- interaction(dd_GT$ID, dd_GT$Run, drop = TRUE)

dd_GT_both <- dd_GT_2nd |> bind_rows(dd_GT[names(dd_GT_2nd)])
```

Prior genotype probabilities assuming Hardy-Weinberg equilibrium:
```{r Defining prior}
allele_counts <- dd_GT_both |> filter(Dilution %in% c(1000, 50), Run %in% c("a", "C")) |> 
  select(Target.ID, A1, A2, Genotype_true_AA) |> 
  group_by(Target.ID) |> 
  summarise(A1 = unique(A1),
            A2 = unique(A2),
            A1A1 = sum(Genotype_true_AA == "A1A1"),
            A2A2 = sum(Genotype_true_AA == "A2A2"),
            A1A2 = sum(Genotype_true_AA == "A1A2")) |> 
  mutate(N1 = 2*A1A1 + A1A2,
         N2 = 2*A2A2 + A1A2)

# We add an offset to the allele counts to avoid 100% major allele frequencies
# and also to have high enough minor allele frequencies that the 21 loci
# with homozygous genotypes in the same allele for all but one sample are
# not consequently wrongly classified because of the prior.
# However, this doesn't seem to be a problem: with an offset of just 1,
# the last chunk in the markdown shows that the wrongly called
# homozygous genotypes don't have extreeme allele frequencies.
offset <- 1
genotype_freq <- allele_counts |> 
  mutate(N1 = N1 + offset, # Add offset to avoid getting 100% allele frequencies
         N2 = N2 + offset) |> 
  mutate(pA = if_else(A1 == "A", N1/(N1+N2), 0), # Allele A is only present as A1
         pC = case_when(A1 == "C" ~ N1/(N1+N2),
                        A2 == "C" ~ N2/(N1+N2),
                        .default = 0),
         pG = case_when(A1 == "G" ~ N1/(N1+N2),
                        A2 == "G" ~ N2/(N1+N2),
                        .default = 0),
         pT = if_else(A2 == "T", N2/(N1+N2), 0)) |> # Allele T is only present as A2
  mutate(AA = pA^2, CC = pC^2, GG = pG^2, TT = pT^2,
         AC = 2*pA*pC, AG = 2*pA*pG, AT = 2*pA*pT,
         CG = 2*pC*pG, CT = 2*pC*pT,
         GT = 2*pG*pT) |> select(Target.ID, AA, CC, GG, TT, AC, AG, AT, CG, CT, GT)
```

# Quick overall comparison per dilution
```{r}
smlr_fit <- dd_GT_both |> filter(Dilution %in% c(50, 31.25)) |> fit_symmetry_model2(intercept = T, b_int = c(0, 1,-2))
dd_GT_both_pred <- dd_GT_both |> predict_prob_threshold(smlr_fit)
dd_pred_dils <- split(dd_GT_both_pred, dd_GT_both_pred$Dilution)
```

```{r}
dd_GT_both_pred <- names(dd_pred_dils) |> lapply(function(x) {
  m <- as.integer(round(2*as.numeric(x)/6))
  e <- 0.02
  dd_pred_dils[[x]] |>
    # predict_mostad_model(m, e, g_prior = "uniform_bi")
    # predict_mostad_model(m, e, g_prior = "uniform_tetra")
    predict_mostad_model(m, e, g_prior = genotype_freq)
}) |> bind_rows()

dd_GT_both_pred |> filter(Dilution >= 31.25) |> 
  summarise(sum(Genotype_pred != Genotype_true)) |> unlist() |> unname()
dd_GT_both_pred |> filter(Dilution >= 31.25) |> 
  summarise(sum(Mostad_pred != Genotype_true)) |> unlist() |> unname()
dd_GT_both_pred |> filter(Dilution >= 31.25) |> 
  summarise(sum(Genotype == no_call)) |> unlist() |> unname()
dd_GT_both_pred |> filter(Dilution >= 31.25) |> 
  summarise(sum(Genotype != no_call & Genotype != Genotype_true)) |> unlist() |> unname()
```

# Final model (after having done the below extensive explorations)
```{r}
dd_mostad_final <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.06, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows()

dd_mostad_bi_final <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.06, "uniform_bi")
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_u = Mostad_pred, Mostad_pmax_u = Mostad_pmax)

dd_mostad_final <- dd_mostad_final |> 
  left_join(dd_mostad_bi_final, by = c("Target.ID", "Dilution", "ID"))


saveRDS(dd_mostad_final, file.path("..", "data", "00_dd_mostad.rds"))
# dd_mostad_final <- readRDS(file.path("..", "data", "00_dd_mostad.rds"))
```


```{r}
dd_mostad_final |> group_by(Dilution) |> 
  summarise(WC_hsg = sum(Genotype_true != Genotype & Genotype != no_call),
            WC_smlr = sum(Genotype_true != Genotype_pred),
            WC_mostad = sum(Genotype_true != Mostad_pred),
            WC_bi = sum(Genotype_true != Mostad_pred_u))
```



# Exploration of Mostad model (led to the above choices)
The chunks below constitutes the exploration of the Mostad model that was done before deciding on the final model above.
We have primarily looked into MLE choices for $e$ and $m$ estimated for each sample, but it turns out that the simple fixed values above give the highest prediction accuracy.

## Estimates of $m$ and $e$ per dilution using all data
The objects `mle_e` and `mle_m` were determined by manually dialing in, i.e., alternating running the next two chunks while manually updating the values in the two objects after each iteration. 
```{r}
mle_e <- c("6.25"=0.003, "12.5"=0.0029, "25"=0.0028, "31.25"=0.0022, "50"=0.0028, "62.5"=0.0027, "125"=0.0027, "250"=0.0027, "500"=0.0027, "1000"=0.0027)
Lm6 <- data.frame("Dilution"=6.25, "m"=2*(2:10), "L"=sapply(2*(2:10), function(y) {predict_mostad_model(dd_pred_dils[["6.25"]], y, mle_e[["6.25"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm12 <- data.frame("Dilution"=12.5, "m"=2*(2:10), "L"=sapply(2*(2:10), function(y) {predict_mostad_model(dd_pred_dils[["12.5"]], y, mle_e[["12.5"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm25 <- data.frame("Dilution"=25, "m"=2*(3:12), "L"=sapply(2*(3:12), function(y) {predict_mostad_model(dd_pred_dils[["25"]], y, mle_e[["25"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm31 <- data.frame("Dilution"=31.25, "m"=2*(5:10), "L"=sapply(2*(5:10), function(y) {predict_mostad_model(dd_pred_dils[["31.25"]], y, mle_e[["31.25"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm50 <- data.frame("Dilution"=50, "m"=2*(6:12), "L"=sapply(2*(6:12), function(y) {predict_mostad_model(dd_pred_dils[["50"]], y, mle_e[["50"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm62 <- data.frame("Dilution"=62.5, "m"=2*(7:15), "L"=sapply(2*(7:15), function(y) {predict_mostad_model(dd_pred_dils[["62.5"]], y, mle_e[["62.5"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm125 <- data.frame("Dilution"=125, "m"=2*(11:20), "L"=sapply(2*(11:20), function(y) {predict_mostad_model(dd_pred_dils[["125"]], y, mle_e[["125"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm250 <- data.frame("Dilution"=250, "m"=2*(21:30), "L"=sapply(2*(21:30), function(y) {predict_mostad_model(dd_pred_dils[["250"]], y, mle_e[["250"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm500 <- data.frame("Dilution"=500, "m"=2*(46:56), "L"=sapply(2*(46:56), function(y) {predict_mostad_model(dd_pred_dils[["500"]], y, mle_e[["500"]], g_prior = genotype_freq, negloglik_me = T)}))
Lm1000 <- data.frame("Dilution"=1000, "m"=2*(58:67), "L"=sapply(2*(58:67), function(y) {predict_mostad_model(dd_pred_dils[["1000"]], y, mle_e[["1000"]], g_prior = genotype_freq, negloglik_me = T)}))
for (dm in list(Lm6, Lm12, Lm25, Lm31, Lm50, Lm62, Lm125, Lm250, Lm500, Lm1000)) {
  m_min <- dm$m[which.min(dm$L)]
  plot(dm$m, dm$L, xlab = "m", ylab = "Likelihood",
       main = paste(dm$Dilution[1], "pg"),
       sub = paste("Minimum at:  m =", m_min, "or", round(m_min/2), "cells. Ought to be:  m =", round(dm$Dilution[1]/6)*2))
}
```

OBS: the DNA amount was only quantified for 1000 pg (with some variance), and for each step down the dilution series, there is also some variance in how well the amount of DNA was halved.
For DNA amounts $31.25\ \text{pg}$ to $62.5\ \text{pg}$, the MLE for $m$ matches the expected number of templates for the DNA amount quite well (within three templates or less than two cells).

At $125\ \text{pg}$ and above, the discrepancy between the MLE and the expected number of templates seems rather big.
However, this isn't expected to alter the Mostad model much, since $m$ enters the model as a parameter determining the proportions $q$ and $1-q$ of the two nucleotides of the true genotype: $q = k/m \sim \text{Binomial}(m, 1/2)$.
Thus, for high $m$, we have $q \approx 1/2$.
That is, the probability function has most of its mass between $0.4$ and $0.6$.
The importance lies in $m$ being captured quite well at lower values where $q$ has a chance of being zero or one, and this does indeed seem to be the case.

The parameter $e$ is related to sequencing or mapping errors in an overall sense.
That is, several mechanisms can affect the value of $e$.
It is tempting to believe that $e$ will depend on the amount of input DNA, such that less DNA leads to higher values for $e$ (which would effectively mean that $e$ depends on $m$), but $e$ will also depend on other factors.
However, if we use the MLEs for $m$ when estimating $e$, we get quite consistent MLEs for $e$ in the range from $0.0027$ to $0.0029$ with a single exception at $31.25\ \text{pg}$ where $e$ becomes $0.0022$.
If we disregard the MLE for $31.25\ \text{pg}$, there might be a slight dependency of $e$ on the amount of DNA, but it is weak enough to be neglected: using a constant value of $e=0.003$ gave the same $m$-estimates as using the MLEs for $e$.
```{r}
mle_m <- c("6.25"=8, "12.5"=10, "25"=12, "31.25"=14, "50"=16, "62.5"=20, "125"=28, "250"=52, "500"=102, "1000"=126)
e_seq <- seq(0.0019, 0.0033, 0.0001)
# mle_m <- c("6.25"=2, "12.5"=4, "25"=8, "31.25"=10, "50"=16, "62.5"=20, "125"=42, "250"=84, "500"=166, "1000"=333)
# e_seq <- seq(0.005, 0.05, 0.005)

Le6 <- data.frame("Dilution"=6.25, "m"=mle_m[["6.25"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["6.25"]], m=mle_m[["6.25"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le12 <- data.frame("Dilution"=12.5, "m"=mle_m[["12.5"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["12.5"]], m=mle_m[["12.5"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le25 <- data.frame("Dilution"=25, "m"=mle_m[["25"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["25"]], m=mle_m[["25"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le31 <- data.frame("Dilution"=31.25, "m"=mle_m[["31.25"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["31.25"]], m=mle_m[["31.25"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le50 <- data.frame("Dilution"=50, "m"=mle_m[["50"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["50"]], m=mle_m[["50"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le62 <- data.frame("Dilution"=62.5, "m"=mle_m[["62.5"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["62.5"]], m=mle_m[["62.5"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le125 <- data.frame("Dilution"=125, "m"=mle_m[["125"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["125"]], m=mle_m[["125"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le250 <- data.frame("Dilution"=250, "m"=mle_m[["250"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["250"]], m=mle_m[["250"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le500 <- data.frame("Dilution"=500, "m"=mle_m[["500"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["500"]], m=mle_m[["500"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
Le1000 <- data.frame("Dilution"=1000, "m"=mle_m[["1000"]], "e"=e_seq, "L"=sapply(e_seq, function(y) {
  predict_mostad_model(dd_pred_dils[["1000"]], m=mle_m[["1000"]], e=y, g_prior = genotype_freq, negloglik_me = T)}))
for (de in list(Le6, Le12, Le25, Le31, Le50, Le62, Le125, Le250, Le500, Le1000)) {
  e_min <- de$e[which.min(de$L)]
  plot(de$e, de$L, xlab = "e", ylab = "Likelihood",
       main = paste(de$Dilution[1], "pg"),
       sub = paste("Minimum at:  e =", e_min))
}
```

# Estimates of $m$ and $e$ per dilution per individual and examination/run
When the Mostad model is applied in practice, then $m$ should in principle be given for each new sample, individually.
This gives us a lot of models, so in order to estimate the parameters efficiently, we only compute the negative log-likelihood in certain points and then apply a loess-regression which can be intrapolated to get the rest of the points needed to determine the MLE. 
```{r}
dd_ind <- split(dd_GT_both_pred, dd_GT_both_pred$Dilution) |> lapply(function(x) {split(x, x$ID)})
```

## Parameter estimates with prior based on data's allele frequencies
```{r}
m_dil_ind <- lapply(names(dd_ind), function(x) {
  dd_dil <- dd_ind[[x]]
  e <- mle_e[[x]]
  if (x == "6.25") {
    m_search <- 2*(1:11)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(4, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:12), 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))

ggplot(m_dil_ind, aes(x=Dilution_f, y=m_est)) + geom_boxplot() + scale_y_continuous(breaks=seq(0, 280, 20))

m_dil_ind |> mutate(m_dil = as.integer(2*round(as.numeric(Dilution)/6))) |> 
ggplot(aes(x=m_dil, y=m_est)) + geom_point() + scale_y_continuous(breaks=seq(0, 280, 20))
```

```{r}
e_dil_ind <- lapply(names(dd_ind), function(x) {
  e_search <- c(seq(0.0002, 0.0008, 0.0002), seq(0.001, 0.004, 0.0005), seq(0.005, 0.007, 0.001))
  
  dd_dil <- dd_ind[[x]]
  e_likelihoods <- lapply(names(dd_dil), function(id) {
    m <- filter(m_dil_ind, Dilution == as.numeric(x), ID == id)$m_est
    sapply(e_search, function(e) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  e_est <- sapply(e_likelihoods, function(i) {
  # for (i in e_likelihoods) {
    dL <- data.frame("e" = e_search, "L" = i)
    Lfit <- loess(L ~ log(e), dL)
    dP <- data.frame("e" = seq(min(e_search), max(e_search), 0.00001))
    dP$Lfit <- predict(Lfit, newdata = dP)
    e_min <- dP$e[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=e)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  e =", e_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = x, e_est = e_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))

ggplot(e_dil_ind, aes(x=Dilution_f, y=e_est)) + geom_boxplot()

e_dil_ind |> mutate(e_dil = as.integer(2*round(as.numeric(Dilution)/6))) |> 
ggplot(aes(x=e_dil, y=e_est)) + geom_point() + 
  geom_point(aes(x=e_dil, y=e_mean), colour="red", shape=3, data = e_dil_ind |> 
               mutate(e_dil = as.integer(2*round(as.numeric(Dilution)/6))) |> 
               group_by(e_dil) |> 
               summarise(e_mean=mean(e_est)))
```

```{r}
m_dil_ind2 <- lapply(names(dd_ind), function(x) {
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- 2*(1:11)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 4, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:12), 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    e <- filter(e_dil_ind, Dilution == as.numeric(x), ID == id)$e_est
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))

ggplot(m_dil_ind, aes(x=Dilution_f, y=m_est)) + geom_boxplot() + scale_y_continuous(breaks=seq(0, 280, 20))

m_dil_ind |> mutate(m_dil = as.integer(2*round(as.numeric(Dilution)/6))) |> 
ggplot(aes(x=m_dil, y=m_est)) + geom_point() + scale_y_continuous(breaks=seq(0, 280, 20))
```

The differences in $m$-estimates between the optimization using a common $e$ for all samples of same dilution and an individual $e$ for each sample and dilution are zero in 164 of 179 cases, and in the 15 non-zero cases, the differences are no greater than 2.
This indicates a stable simultaneous optimum for both parameters.
```{r}
m_comp <- m_dil_ind[,1:3] |> left_join(m_dil_ind2[,1:3], by=c("ID", "Dilution"))
nrow(m_comp)
sum(m_comp$m_est.x == m_comp$m_est.y)
m_comp[m_comp$m_est.x != m_comp$m_est.y,]
```

```{r}
e_dil_ind2 <- lapply(names(dd_ind), function(x) {
  dd_dil <- dd_ind[[x]]
  e_search <- c(seq(0.0002, 0.0008, 0.0002), seq(0.001, 0.004, 0.0005), seq(0.005, 0.007, 0.001))
  
  e_likelihoods <- lapply(names(dd_dil), function(id) {
    m <- filter(m_dil_ind2, Dilution == as.numeric(x), ID == id)$m_est
    sapply(e_search, function(e) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  e_est <- sapply(e_likelihoods, function(i) {
  # for (i in e_likelihoods) {
    dL <- data.frame("e" = e_search, "L" = i)
    Lfit <- loess(L ~ log(e), dL)
    dP <- data.frame("e" = seq(min(e_search), max(e_search), 0.00001))
    dP$Lfit <- predict(Lfit, newdata = dP)
    e_min <- dP$e[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=e)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  e =", e_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = x, e_est = e_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))

ggplot(e_dil_ind2, aes(x=Dilution_f, y=e_est)) + geom_boxplot()

e_dil_ind2 |> mutate(e_dil = as.integer(2*round(as.numeric(Dilution)/6))) |> 
ggplot(aes(x=e_dil, y=e_est)) + geom_point() + 
  geom_point(aes(x=e_dil, y=e_mean), colour="red", shape=3, data = e_dil_ind |> 
               mutate(e_dil = as.integer(2*round(as.numeric(Dilution)/6))) |> 
               group_by(e_dil) |> 
               summarise(e_mean=mean(e_est)))
```

The differences between the first and second optimization of $e$ using the most recent individually estimated $m$ for each sample and dilution are zero in 175 of 179 cases, and in the 4 non-zero cases, the differences are at max 0.00004.
This indicates a stable simultaneous optimum for both parameters.
```{r}
e_comp <- e_dil_ind[,1:3] |> left_join(e_dil_ind2[,1:3], by=c("ID", "Dilution"))
sum(e_comp$e_est.x == e_comp$e_est.y)
e_comp[e_comp$e_est.x != e_comp$e_est.y,]
```


## Parameter estimates with biallelic uniform prior
```{r}
m_dil_ind_bi <- lapply(names(dd_ind), function(x) {
  dd_dil <- dd_ind[[x]]
  e <- mle_e[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(4, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:12), 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 16, 20, 26, 32, 36, 40, 46, 50, 52, 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = "uniform_bi", negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```

```{r}
e_dil_ind_bi <- lapply(names(dd_ind), function(x) {
  e_search <- c(seq(0.0002, 0.0008, 0.0002), seq(0.001, 0.004, 0.0005), seq(0.005, 0.007, 0.001))
  
  dd_dil <- dd_ind[[x]]
  e_likelihoods <- lapply(names(dd_dil), function(id) {
    m <- filter(m_dil_ind_bi, Dilution == as.numeric(x), ID == id)$m_est
    sapply(e_search, function(e) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = "uniform_bi", negloglik_me = T)
    })
  })
  e_est <- sapply(e_likelihoods, function(i) {
  # for (i in e_likelihoods) {
    dL <- data.frame("e" = e_search, "L" = i)
    Lfit <- loess(L ~ log(e), dL)
    dP <- data.frame("e" = seq(min(e_search), max(e_search), 0.00001))
    dP$Lfit <- predict(Lfit, newdata = dP)
    e_min <- dP$e[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=e)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  e =", e_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = x, e_est = e_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```

```{r}
m_dil_ind2_bi <- lapply(names(dd_ind), function(x) {
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 4, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:12), 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 16, 20, 26, 32, 36, 40, 46, 50, 52, 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    e <- filter(e_dil_ind_bi, Dilution == as.numeric(x), ID == id)$e_est
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = "uniform_bi", negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```

The differences in $m$-estimates between the optimization using a common $e$ for all samples of same dilution and an individual $e$ for each sample and dilution are zero in 165 of 179 cases, and in the 14 non-zero cases, the differences are no greater than 2.
This indicates a stable simultaneous optimum for both parameters.
```{r}
m_comp <- m_dil_ind_bi[,1:3] |> left_join(m_dil_ind2_bi[,1:3], by=c("ID", "Dilution"))
sum(m_comp$m_est.x == m_comp$m_est.y)
m_comp[m_comp$m_est.x != m_comp$m_est.y,]
```

```{r}
e_dil_ind2_bi <- lapply(names(dd_ind), function(x) {
  dd_dil <- dd_ind[[x]]
  e_search <- c(seq(0.0002, 0.0008, 0.0002), seq(0.001, 0.004, 0.0005), seq(0.005, 0.007, 0.001))
  
  e_likelihoods <- lapply(names(dd_dil), function(id) {
    m <- filter(m_dil_ind2_bi, Dilution == as.numeric(x), ID == id)$m_est
    sapply(e_search, function(e) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = "uniform_bi", negloglik_me = T)
    })
  })
  e_est <- sapply(e_likelihoods, function(i) {
  # for (i in e_likelihoods) {
    dL <- data.frame("e" = e_search, "L" = i)
    Lfit <- loess(L ~ log(e), dL)
    dP <- data.frame("e" = seq(min(e_search), max(e_search), 0.00001))
    dP$Lfit <- predict(Lfit, newdata = dP)
    e_min <- dP$e[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=e)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  e =", e_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = x, e_est = e_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```

The differences between the first and second optimization of $e$ using the most recent individually estimated $m$ for each sample and dilution are zero in 173 of 179 cases, and in the 6 non-zero cases, the differences are at max 0.00008.
This indicates a stable simultaneous optimum for both parameters.
```{r}
e_comp <- e_dil_ind_bi[,1:3] |> left_join(e_dil_ind2_bi[,1:3], by=c("ID", "Dilution"))
sum(e_comp$e_est.x == e_comp$e_est.y)
e_comp[e_comp$e_est.x != e_comp$e_est.y,]
```

### How well do the rather low MLEs for $e$ match the proportion of false counts in the data?
Mostad et al. writes:
"For homozygous loci, our observational model predicts that each read has an independent probability 3e/4 of being of a type not present in the genotype; for heterozygous loci the probability is 2e/4."
An empirical estimate of $e$ could then be found as $e_{emp} = 4\times\frac{\text{false rate}}{3}$ for homozygous loci and $e_{emp} = 4\times\frac{\text{false rate}}{2}$ for heterozygous loci.
Below we see that the empirical versions of $e$ are much higher than the MLEs.
This indicates that the false rate is not well modelled by the MLEs for $e$.
Later, we will see that models using the MLEs for $e$ have lower prediction accuracy than when a higher $e$ is used.
Higher values than the MLEs should be used, but while it seems most appropriate to let $e$ depend on the dilution (amount of DNA and thus implicit dependence on $m$), we will later see that using $e=0.06$ for all dilutions actually gives a better classification accuracy.
```{r}
false_counts <- dd_GT_both |> 
  mutate(false_rate = case_when(Genotype_true_AA == "A1A1" ~ (Coverage-A1.Reads)/Coverage,
                                Genotype_true_AA == "A2A2" ~ (Coverage-A2.Reads)/Coverage,
                                .default = (Coverage-A1.Reads-A2.Reads)/Coverage),
         e_emp = if_else(Genotype_true_AA == "A1A2", false_rate*4/2, false_rate*4/3))

(e_emp0 <- false_counts |> group_by(Dilution) |>
  summarise("Min" = min(e_emp),
            "Median" = median(e_emp),
            "Mean" = mean(e_emp),
            "Q75" = quantile(e_emp, probs=0.75),
            "Q90" = quantile(e_emp, probs=0.9),
            "Q95" = quantile(e_emp, probs=0.95),
            "Q99" = quantile(e_emp, probs=0.99)))

(e_emp <- false_counts |> filter(false_rate > 0) |> group_by(Dilution) |>
  summarise("Min" = min(e_emp),
            "Median" = median(e_emp),
            "Mean" = mean(e_emp),
            "Q75" = quantile(e_emp, probs=0.75),
            "Q90" = quantile(e_emp, probs=0.9),
            "Q95" = quantile(e_emp, probs=0.95),
            "Q96" = quantile(e_emp, probs=0.96),
            "Q97" = quantile(e_emp, probs=0.97),
            "Q98" = quantile(e_emp, probs=0.98),
            "Q99" = quantile(e_emp, probs=0.99)))
```


## Parameter estimate of $m$ with $e=0.01$
```{r}
m_dil_ind_01 <- lapply(names(dd_ind), function(x) {
  e <- 0.01
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.015$
```{r}
m_dil_ind_015 <- lapply(names(dd_ind), function(x) {
  e <- 0.015
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.018$
```{r}
m_dil_ind_018 <- lapply(names(dd_ind), function(x) {
  e <- 0.018
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.02$
```{r}
m_dil_ind_02 <- lapply(names(dd_ind), function(x) {
  e <- 0.02
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.03$
```{r}
m_dil_ind_03 <- lapply(names(dd_ind), function(x) {
  e <- 0.03
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.04$
```{r}
m_dil_ind_04 <- lapply(names(dd_ind), function(x) {
  e <- 0.04
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.05$
```{r}
m_dil_ind_05 <- lapply(names(dd_ind), function(x) {
  e <- 0.05
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.06$
```{r}
m_dil_ind_06 <- lapply(names(dd_ind), function(x) {
  e <- 0.06
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.07$
```{r}
m_dil_ind_07 <- lapply(names(dd_ind), function(x) {
  e <- 0.07
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.1$
```{r}
m_dil_ind_1 <- lapply(names(dd_ind), function(x) {
  e <- 0.1
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.11$
Only relevant for 6.25, 12.5, 25, 50, and 62.5 pg
```{r}
m_dil_ind_11 <- lapply(names(dd_ind)[c(1:3, 5:6)], function(x) {
  e <- 0.11
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.12$
Only relevant for 6.25, 12.5, 25, 50, and 62.5 pg
```{r}
m_dil_ind_12 <- lapply(names(dd_ind)[c(1:3, 5:6)], function(x) {
  e <- 0.12
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.13$
Only relevant for 6.25, 12.5, 25, 50, and 62.5 pg
```{r}
m_dil_ind_13 <- lapply(names(dd_ind)[c(1:3, 5:6)], function(x) {
  e <- 0.13
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.14$
The log-likelihood experiences numerical overflow for one of the individuals in the data subset of 1000 pg DNA when $e > 0.146$.
Therefore, we test $e=14$ rather than $e=0.15$.
```{r}
m_dil_ind_14 <- lapply(names(dd_ind), function(x) {
  e <- 0.14
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.15$
Only relevant for 6.25 pg
```{r}
m_dil_ind_15 <- lapply("6.25", function(x) {
  e <- 0.15
  dd_dil <- dd_ind[[x]]
  m_search <- c(2, 4, 8, 12, 16, 20, 26)
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


## Parameter estimate of $m$ with $e=0.2$
The log-likelihood experiences numerical overflow for more of the individuals in the data subset of 500 and 1000 pg DNA when $e = 0.2$.
We simply filter away the 500 and 1000 pg dilution when exploring the higher values of $e$.
```{r}
m_dil_ind_2 <- lapply(names(dd_ind)[-c(9:10)], function(x) {
  e <- 0.2
  dd_dil <- dd_ind[[x]]
  if (x == "6.25") {
    m_search <- c(2, 4, 8, 12, 16, 20, 26)
  } else if (x == "12.5") {
    m_search <- 2*(1:11)
  } else if (x == "25") {
    m_search <- 2*(1:13)
  } else if (x == "31.25") {
    m_search <- c(2, 2*(4:8), 24, 28, 34, 40)
  } else if (x == "50") {
    m_search <- c(4, 2*(6:9), 24, 30, 36, 42)
  } else if (x == "62.5") {
    m_search <- c(6, 2*(8:15), 36, 42)
  } else if (x == "125") {
    m_search <- c(8, 2*(17:26), 58, 64)
  } else if (x == "250") {
    m_search <- c(8, 38, 44, 46, 50, 62, 72, 76, 96, 98, 112, 114, 116, 120, 130)
  } else if (x == "500") {
    m_search <- c(16, 94, 102, 112, 122, 144, 150, 186, 194, 200)
  } else if (x == "1000") {
    m_search <- c(16, 96, 108, 120, 132, 144, 160, 164, 214, 216, 240, 242, 292, 296, 300)
  }
  m_likelihoods <- lapply(names(dd_dil), function(id) {
    sapply(m_search, function(m) {
      predict_mostad_model(dd_dil[[id]], m, e, g_prior = genotype_freq, negloglik_me = T)
    })
  })
  m_est <- sapply(m_likelihoods, function(i) {
  # for (i in m_likelihoods) {
    dL <- data.frame("m" = m_search, "L" = i)
    Lfit <- loess(L ~ log(m), dL)
    dP <- data.frame("m" = min(m_search):max(m_search))
    dP$Lfit <- predict(Lfit, newdata = dP)
    m_min <- dP$m[which.min(dP$Lfit)]
    # p <- ggplot(dL, aes(x=m)) +
    #   geom_point(aes(y=L)) +
    #   geom_line(aes(y=Lfit), data = dP) +
    #   labs(title = paste("Loess minimum at:  m =", m_min))
    # print(p)}
  })
  data.frame(ID = names(dd_dil), Dilution = as.numeric(x), m_est = m_est)
}) |> bind_rows() |>
  mutate(Dilution_f = factor(Dilution, levels = c(1000, 500, 250, 125, 62.5, 50, 31.25, 25, 12.5)))
```


# Model predictions
The model from Mostad et al. relies on the integer parameter $m$:
"representing the approximate number of DNA templates from the sample that end up founding PCR amplicons for this locus".
This means that the model gets tuned with information about the dilution level.
The way the model is defined with $q = k/m$ and $k \sim \text{Binomial}(m, 1/2)$ suggests that the quote, "representing the approximate number of DNA templates", should be taken literally, i.e., that $m$ is two times the number of cells in each level of the dilution series.
For the dilution level with $31.25\ \text{pg DNA}$, one could then set $m = \frac{31.25\ \text{[pg]}}{6\ \text{[pg/cell]}} \times 2\ \text{[templates/cell]} \approx 10\ \text{[templates]}$.
However, since we already introduced a prior $P(g)$ in order to derive $P(g \mid c) = \frac{P(c \mid g)}{P(c)} P(g)$, we have actually already derived a likelihood for $m$ and $e$ where $g$ has been marginalized away:
$$P(c) = P(c \mid m, e) = L(m, e \mid c).$$

In the cross-validation, we don't want to introduce a prior into the SMLR model, so it is fitted on the train data and evaluated on the test data, but because of this template/sample dependent $m$-parameter, this is not very meaningful for the observational model from Mostad et al.
It seems more meaningful to determine $m$ for the sample the model is used on, either by quantifying the DNA before PCR or by maximizing $L(m, e \mid c)$.
The overall error parameter, $e$, is likely to be lab dependent, but it may also be affected by sample dependent factors.
Thus, one should either already have determined $e$ for the lab setup before applying the model to new data, or one should again maximize $L(m, e \mid c)$.
In a few chunks ahead, we evaluated the models' number of WCs (within each dilutions) in different ways (manually out-commenting the relevant lines):

- with the $e$ and $m$ found above per individual by maximum likelihood,
  + this gave very low $e$-parameters, since most observations comes with no or just a few false counts, but such low $e$ don't capture the true false count rate well for the $10\%$ of observations with false counts above the "just a few counts" - with too low $e$, the model will take the false counts for the before mentioned $10\%$ as strong evidence of an allele, also for cases where the true allele signal is over $4.000\%$ higher,
  + therefore, we also tried out $e$-values in the range $[0.01; 0.14]$.
  + the MLEs for $m$ varies rather much, sometimes with large deviance away from what would be expected of the dilution,
- with $e \in [0.01; 0.14]$ and $m$ found above per individual by maximum likelihood,
  + the MLEs for $m$ varies rather much, sometimes with large deviance away from what would be expected of the dilution,
- with $e \in [0.01; 0.14]$ and $m$ determined as $m = \text{round}\left(\frac{\text{DNA amount}\ \text{[pg]}}{6\ \text{[pg/cell]}} \times 2\ \text{[templates/cell]}\right) \approx \text{Number}\ \text{[templates]}$.

It turns out that $e \in [0.01; 0.06]$ generally gives fewer WCs (practically independent of the dilution).
It also turns out that the model isn't very prone to whether $m$ is the MLE or the value derived from the DNA amount - in fact, the latter seems to give slightly fewer WCs, or at least the same amount of WCs but for a wider range of $e$-values.

All in all, the Mostad model generally seems to produce the fewest WCs by choosing $e = 0.06$ and setting $m$ to the value expected from the amount of DNA.
This suggest a static model, so cross-validation would only make sense in order to evaluate the models sensitivity to the choice of prior.
However, using a uniform biallelic prior will also help in such assessment, and it is easier to defend, since we should else argue how much the prior should be allowed to vary during a cross-validation.
Therefore, the static Mostad model is evaluated directly on all samples within each dilution (configuration of $m$), and the results are appended to Table 1.

### Assessment of the number of WCs produces by different configurations of the Mostad model.
```{r}
# The original ID-variable in the first dilution series has been
# overwritten with interaction(ID, Run), such that the same individual
# got new IDs for each run/examination.
dd_mostad <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      m <- filter(m_dil_ind2, Dilution == dil, ID == id)$m_est
      e <- filter(e_dil_ind2, Dilution == dil, ID == id)$e_est
      predict_mostad_model(x, m, e, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows()

dd_mostad_bi <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      m <- filter(m_dil_ind2, Dilution == dil, ID == id)$m_est
      e <- filter(e_dil_ind2, Dilution == dil, ID == id)$e_est
      predict_mostad_model(x, m, e, "uniform_bi")
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_bi = Mostad_pred, Mostad_pmax_bi = Mostad_pmax)

dd_mostad_01 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_01, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.01, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_01 = Mostad_pred, Mostad_pmax_01 = Mostad_pmax)

dd_mostad_015 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_015, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.015, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_015 = Mostad_pred, Mostad_pmax_015 = Mostad_pmax)

dd_mostad_018 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_018, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.018, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_018 = Mostad_pred, Mostad_pmax_018 = Mostad_pmax)

dd_mostad_02 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_02, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.02, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_02 = Mostad_pred, Mostad_pmax_02 = Mostad_pmax)

dd_mostad_03 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_03, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.03, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_03 = Mostad_pred, Mostad_pmax_03 = Mostad_pmax)

dd_mostad_04 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_04, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.04, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_04 = Mostad_pred, Mostad_pmax_04 = Mostad_pmax)

dd_mostad_05 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_05, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.05, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_05 = Mostad_pred, Mostad_pmax_05 = Mostad_pmax)

dd_mostad_06 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_06, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.06, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_06 = Mostad_pred, Mostad_pmax_06 = Mostad_pmax)

dd_mostad_07 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_07, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.07, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_07 = Mostad_pred, Mostad_pmax_07 = Mostad_pmax)

dd_mostad_10 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_1, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.1, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_10 = Mostad_pred, Mostad_pmax_10 = Mostad_pmax)

dd_mostad_11 <- names(dd_pred_dils)[c(1:3, 5:6)] |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_11, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.11, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_11 = Mostad_pred, Mostad_pmax_11 = Mostad_pmax)

dd_mostad_12 <- names(dd_pred_dils)[c(1:3, 5:6)] |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_12, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.12, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_12 = Mostad_pred, Mostad_pmax_12 = Mostad_pmax)

dd_mostad_13 <- names(dd_pred_dils)[c(1:3, 5:6)] |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_13, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.13, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_13 = Mostad_pred, Mostad_pmax_13 = Mostad_pmax)

dd_mostad_14 <- names(dd_pred_dils) |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_14, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.14, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_14 = Mostad_pred, Mostad_pmax_14 = Mostad_pmax)

dd_mostad_15 <- "6.25" |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_15, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.15, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_15 = Mostad_pred, Mostad_pmax_15 = Mostad_pmax)

dd_mostad_20 <- names(dd_pred_dils)[-c(9:10)] |> lapply(function(dil) {
  split(dd_pred_dils[[dil]], dd_pred_dils[[dil]]$ID) |> 
    lapply(function(x) {
      id <- unique(x$ID)
      # m <- filter(m_dil_ind_2, Dilution == dil, ID == id)$m_est
      m <- round(2*as.numeric(dil)/6)
      predict_mostad_model(x, m, e=0.2, genotype_freq)
    }) |> bind_rows()
}) |> bind_rows() |> select(Target.ID, Dilution, ID, Mostad_pred_20 = Mostad_pred, Mostad_pmax_20 = Mostad_pmax)

dd_mostad_all <- dd_mostad |> 
  left_join(dd_mostad_bi, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_01, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_015, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_018, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_02, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_03, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_04, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_05, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_06, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_07, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_10, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_11, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_12, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_13, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_14, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_15, by = c("Target.ID", "Dilution", "ID")) |> 
  left_join(dd_mostad_20, by = c("Target.ID", "Dilution", "ID"))
```

We will explore how many WCs the different $e$ parameters make within each dilution.
It looks like $e \in [0.02; 0.14]$ gives the fewest WCs for different dillutions.
```{r}
dd_mostad_all |> group_by(Dilution) |> 
  summarise(WC_smlr = sum(Genotype_true != Genotype_pred),
            WC_gf = sum(Genotype_true != Mostad_pred),
            WC_bi = sum(Genotype_true != Mostad_pred_bi),
            # WC_01 = sum(Genotype_true != Mostad_pred_01),
            # WC_015 = sum(Genotype_true != Mostad_pred_015),
            # WC_018 = sum(Genotype_true != Mostad_pred_018),
            WC_02 = sum(Genotype_true != Mostad_pred_02),
            WC_03 = sum(Genotype_true != Mostad_pred_03),
            WC_04 = sum(Genotype_true != Mostad_pred_04),
            WC_05 = sum(Genotype_true != Mostad_pred_05),
            WC_06 = sum(Genotype_true != Mostad_pred_06),
            WC_07 = sum(Genotype_true != Mostad_pred_07),
            WC_10 = sum(Genotype_true != Mostad_pred_10),
            WC_11 = sum(Genotype_true != Mostad_pred_11),
            WC_12 = sum(Genotype_true != Mostad_pred_12),
            WC_13 = sum(Genotype_true != Mostad_pred_13),
            WC_14 = sum(Genotype_true != Mostad_pred_14),
            WC_15 = sum(Genotype_true != Mostad_pred_15),
            WC_20 = sum(Genotype_true != Mostad_pred_20))
```


As seen below, even with an offset of just 1 (in the creation of the prior), it doesn't look like that the wrongly called homozygous genotypes are wrong because of extreme allele frequencies used to create the prior: the allele counts for these homozygous wrong calls are not at all extreme.
But we note that the Mostad model sometimes estimate a heterozygous genotype probability so high that it has been rounded to one even though the ratio $\frac{s_{min}}{s_{max}}$ was very low, so apparently the Mostad model sees even a few reads as high evidence for an actual allele present: this is likely because of the very low $e$-estimates.
Therefore, we also tested a model with $e=0.02$ as used in Mostad et al. which actually had fewer wrong calls, but the model sometimes still estimated a high heterozygous genotype probability despite a low ratio of $\frac{s_{min}}{s_{max}}$.
```{r}
mostad_wc <- dd_mostad_all |> 
  filter(Mostad_pred != Genotype_true, Dilution >= 31.25) |> 
  left_join(allele_counts[,c("Target.ID", "N1", "N2")], by = "Target.ID")
mostad_wc <- dd_mostad_all |> 
  filter(Mostad_pred_06 != Genotype_true, Dilution >= 31.25) |> 
  left_join(allele_counts[,c("Target.ID", "N1", "N2")], by = "Target.ID")
```