---
title: "Bootstrap"
author: "Corresponding author"
date: "2023-12-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggnewscale) # Enables multiple colour- and fill-scales
library(latex2exp)
library(patchwork)
source("00-functions-and-global-definitions.R")

plot_title_size <- 7
axis_title_size <- 7
axis_ticks_size <- 7
legend_title_size <- 7
legend_text_size <- 7
```

```{r Reading data}
dd_GT <- readRDS(file.path("..", "data", "00_dd_GT.rds"))
dd_GT_2nd <- readRDS(file.path("..", "data", "00_dd_GT_2nd.rds"))
true_profiles_2nd <- readRDS(file.path("..", "data", "00_true_profiles_2nd.rds"))

dd_GT_both <- dd_GT_2nd |> bind_rows(dd_GT[names(dd_GT_2nd)])
```


# Bootstrap
What is the proportion of heterozygous genotypes in the different dilutions?
```{r}
dd_GT |> group_by(Dilution, Run) |> 
  summarise("Proportion a1a1" = sum(Genotype_true_AA == "A1A1")/n(),
            "Proportion a2a2" = sum(Genotype_true_AA == "A2A2")/n(),
            "Proportion a1a2" = sum(Genotype_true_AA == "A1A2")/n())

dd_GT_2nd |> group_by(Dilution) |> 
  summarise("Proportion a1a1" = sum(Genotype_true_AA == "A1A1")/n(),
            "Proportion a2a2" = sum(Genotype_true_AA == "A2A2")/n(),
            "Proportion a1a2" = sum(Genotype_true_AA == "A1A2")/n())
```


```{r include=FALSE}
dils <- list(50, c(50, 25), 25, 12.5, 6.25)
init_model_dil_fine <- lapply(dils, function(d) {
  df <- dd_GT_2nd |> filter(Dilution %in% d)
  m <- fit_symmetry_model2(df, f = function(x) sqrt(x), intercept = TRUE, control_list = list(maxit = 10^3))
  m$Dilution <- paste0(d, "pg", collapse = " & ")
  m$Het_prop_pop <- nrow(filter(true_profiles_2nd, Genotype_true_AA == "A1A2")) / nrow(true_profiles_2nd)
  m$f <- "sqrt"
  m$beta0 <- m$par[1]
  m$beta1 <- m$par[2]
  m$beta2 <- m$par[3]
  m$Slope <- -m$beta1/m$beta2
  m$Intercept_1 <- -m$beta0/m$beta2
  m$Intercept_2 <- -m$beta0/m$beta1
  return(m[-1])
}) |> bind_rows() |>
  pivot_longer(cols = c("beta0", "beta1", "beta2", "Slope", "Intercept_1", "Intercept_2"),
               names_to = "Parameter",
               values_to = "Sample_estimate")

dd_boot_fine <- readRDS(file.path("..", "data", "01_obj_bootstrap_sqrt_1K.rds"))

dd_boot_fine <- dd_boot_fine |> 
  mutate("Complete separation" = if_else(n_WC == 0, "Yes", "No")) |> 
  arrange(`Complete separation`) |>
  group_by(Dilution, Samplesize) |> 
  mutate("Slope" = -beta1/beta2,
         "Intercept_1" = -beta0/beta2,
         "Intercept_2" = -beta0/beta1)

dd_boot_long_fine <- dd_boot_fine |>
  pivot_longer(cols = c("beta0", "beta1", "beta2", "Slope", "Intercept_1", "Intercept_2"),
               names_to = "Parameter",
               values_to = "Estimate") |> 
  filter(Samplesize >= 4)
```

# Value of the negative log-likelihood function at the found set of parameters
```{r}
dd_boot_fine |> filter(Dilution=="50pg", Samplesize == 1) |> 
  arrange(Value) |> 
  mutate(Idx=1:n()) |> 
ggplot(aes(x=`Complete separation`, y=Value, colour=`Complete separation`)) + 
  geom_point() +
  ylab("Negative log-likelihood")
```


```{r echo=FALSE}
quantiles <- c(0.1, 0.9)
p1 <- plot_boot(dd_boot_long_fine, "beta0", "50pg", Y_lim=F, Q=quantiles, full_fit=init_model_dil_fine, y_lab = TeX("$\\beta_0$"), main_title = "50pg", add_legend = T)
p1_2 <- plot_boot(dd_boot_long_fine, "beta0", "50pg & 25pg", Y_lim=c(-0.5,2.5), Q=quantiles, full_fit=init_model_dil_fine, main_title = "50pg & 25pg")
p2 <- plot_boot(dd_boot_long_fine, "beta0", "25pg", Y_lim=c(-0.5,2.5), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine, main_title = "25pg")
p3 <- plot_boot(dd_boot_long_fine, "beta0", "12.5pg", Y_lim=c(-0.5,2.5), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine, main_title = "12.5pg")
p4 <- plot_boot(dd_boot_long_fine, "beta0", "6.25pg", Y_lim=c(-0.5,2.5), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine, main_title = "6.25pg")

p5 <- plot_boot(dd_boot_long_fine, "beta1", "50pg", Y_lim=F, Q=quantiles, full_fit=init_model_dil_fine, y_lab = TeX("$\\beta_1$"))
p5_6 <- plot_boot(dd_boot_long_fine, "beta1", "50pg & 25pg", Y_lim=c(0.1,0.7), Q=quantiles, full_fit=init_model_dil_fine)
p6 <- plot_boot(dd_boot_long_fine, "beta1", "25pg", Y_lim=c(0.1,0.7), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p7 <- plot_boot(dd_boot_long_fine, "beta1", "12.5pg", Y_lim=c(0.1,0.7), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p8 <- plot_boot(dd_boot_long_fine, "beta1", "6.25pg", Y_lim=c(0.1,0.7), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)

p9 <- plot_boot(dd_boot_long_fine, "beta2", "50pg", Y_lim=F, Q=quantiles, full_fit=init_model_dil_fine, y_lab = TeX("$\\beta_2$"))
p9_10 <- plot_boot(dd_boot_long_fine, "beta2", "50pg & 25pg", Y_lim=c(-2.5,-0.9), Q=quantiles, full_fit=init_model_dil_fine)
p10 <- plot_boot(dd_boot_long_fine, "beta2", "25pg", Y_lim=c(-2.5,-0.9), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p11 <- plot_boot(dd_boot_long_fine, "beta2", "12.5pg", Y_lim=c(-2.5,-0.9), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p12 <- plot_boot(dd_boot_long_fine, "beta2", "6.25pg", Y_lim=c(-2.5,-0.9), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)

p13 <- plot_boot(dd_boot_long_fine, "Intercept_1", "50pg", Y_lim=F, Q=quantiles, full_fit=init_model_dil_fine, y_lab = TeX("$-\\beta_0/\\beta_2$"))
p13_14 <- plot_boot(dd_boot_long_fine, "Intercept_1", "50pg & 25pg", Y_lim=c(-0.5,1.5), Q=quantiles, full_fit=init_model_dil_fine)
p14 <- plot_boot(dd_boot_long_fine, "Intercept_1", "25pg", Y_lim=c(-0.5,1.5), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p15 <- plot_boot(dd_boot_long_fine, "Intercept_1", "12.5pg", Y_lim=c(-0.5,1.5), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p16 <- plot_boot(dd_boot_long_fine, "Intercept_1", "6.25pg", Y_lim=c(-0.5,1.5), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)

p13_2 <- plot_boot(dd_boot_long_fine, "Intercept_2", "50pg", Y_lim=F, Q=quantiles, full_fit=init_model_dil_fine, y_lab = TeX("$-\\beta_0/\\beta_1$"))
p13_14_2 <- plot_boot(dd_boot_long_fine, "Intercept_2", "50pg & 25pg", Y_lim=c(-25,1), Q=quantiles, full_fit=init_model_dil_fine)
p14_2 <- plot_boot(dd_boot_long_fine, "Intercept_2", "25pg", Y_lim=c(-25,1), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p15_2 <- plot_boot(dd_boot_long_fine, "Intercept_2", "12.5pg", Y_lim=c(-25,1), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)
p16_2 <- plot_boot(dd_boot_long_fine, "Intercept_2", "6.25pg", Y_lim=c(-25,1), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine)

p17 <- plot_boot(dd_boot_long_fine, "Slope", "50pg", Y_lim=F, Q=quantiles, full_fit=init_model_dil_fine, y_lab = TeX("$-\\beta_1/\\beta_2$"), x_tick = T)
p17_18 <- plot_boot(dd_boot_long_fine, "Slope", "50pg & 25pg", Y_lim=c(0,0.4), Q=quantiles, full_fit=init_model_dil_fine, x_tick = T)
p18 <- plot_boot(dd_boot_long_fine, "Slope", "25pg", Y_lim=c(0,0.4), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine, x_tick = T, x_lab = "Sample size (162 SNPs per individual per DNA quantity)")
p19 <- plot_boot(dd_boot_long_fine, "Slope", "12.5pg", Y_lim=c(0,0.4), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine, x_tick = T)
p20 <- plot_boot(dd_boot_long_fine, "Slope", "6.25pg", Y_lim=c(0,0.4), y_tick=F, Q=quantiles, full_fit=init_model_dil_fine, x_tick = T)

p <- p1 + p1_2 + p2 + p3 + p4 +
     p5 + p5_6 + p6 + p7 + p8 +
     p9 + p9_10 + p10 + p11 + p12 +
     p13 + p13_14 + p14 + p15 + p16 +
     p13_2 + p13_14_2 + p14_2 + p15_2 + p16_2 +
     p17 + p17_18 + p18 + p19 + p20 +
     plot_layout(ncol = 5, nrow = 6, byrow = T, guides = "collect", axis_titles = "collect_x")

# p <- p1 + p1_2 + p2 + p3 + 
#      p5 + p5_6 + p6 + p7 + 
#      p9 + p9_10 + p10 + p11 + 
#      p13 + p13_14 + p14 + p15 + 
#      p17 + p17_18 + p18 + p19 + 
#      plot_layout(ncol = 4, nrow = 5, byrow = T, guides = "collect")
for (i in seq_along(p)) {
  if (i%%5 == 1) { # leftmost plots
    p[[i]]$theme$plot.margin[4] <- 0*p[[i]]$theme$plot.margin[4]
  } else if (i%%5 == 0) { # rightmost plots
    p[[i]]$theme$plot.margin[2] <- 0*p[[i]]$theme$plot.margin[2]
  }
  if (i <= 5) { # top row of plots
    p[[i]]$theme$plot.margin[1] <- 0*p[[i]]$theme$plot.margin[1]
  } else if (i > 25) { # bottom row of plots
    p[[i]]$theme$plot.margin[3] <- 0*p[[i]]$theme$plot.margin[3]
  }
  # All plots should have their legend's right margin set to zero
  p[[i]]$theme$legend.margin[2] <- 0*p[[i]]$theme$legend.margin[2]
}
# The overall margins added by patchwork is set to zero:
p <- p + plot_annotation(theme = theme(plot.margin = margin(t=0, r=0, b=0, l=0, unit = 'pt')))

ggsave(filename = file.path("..", "article-FSIGEN", "FIG_S1-two-column.bmp"),
       plot = p, device = "bmp", width = 7480, height = 4330, units = "px", dpi = 1000)
```


Note: The data contains around $3/4$ homozygous genotypes and $1/4$ heterozygous genotypes, but the majority of the wrong calls are found among the heterozygous genotypes. A heterozygous genotype is generally more prone to become a wrong call, since one of its alleles can drop out leaving only the signal from the other allele (while an allele drop out for a homozygous genotype still leaves an intact signal from the other allele which is of the same kind).
```{r}
dils2 <- unique(dd_GT_both$Dilution)[order(unique(dd_GT_both$Dilution), decreasing = T)]
model_dil <- lapply(dils2, function(d) {
  df <- dd_GT_both |> filter(Dilution %in% d)
  m <- fit_symmetry_model2(df, f = function(x) sqrt(x), intercept = TRUE, method = "Nelder-Mead", control_list = list(maxit = 500))
  df <- predict_prob_threshold(df, m)
  return(df)
}) |> bind_rows()
# names(model_dil) <- paste0(dils2, "pg")

model_dil |> group_by(Dilution) |> 
  summarise("%Het" = round(sum(Genotype_true_AA == "A1A2")/n()*100, digits = 1),
            "%Hom" = round(sum(Genotype_true_AA != "A1A2")/n()*100, digits = 1),
            "%WC" = round(sum(Genotype_true != Genotype_pred)/n()*100, digits = 2),
            "#WC|Het" = sum((Genotype_true != Genotype_pred) & (Genotype_true_AA == "A1A2")),
            "#WC|Hom" = sum((Genotype_true != Genotype_pred) & (Genotype_true_AA != "A1A2")),
            "%WC among Het" = round(sum((Genotype_true != Genotype_pred) & (Genotype_true_AA == "A1A2"))*100/sum(Genotype_true_AA == "A1A2"), digits = 2),
            "%WC among Hom" = round(sum((Genotype_true != Genotype_pred) & (Genotype_true_AA != "A1A2"))*100/sum(Genotype_true_AA != "A1A2"), digits = 2)) |> 
  arrange(desc(Dilution))
```


# Elbow in standard deviation across sample size
```{r}
param <- c("beta0", "beta1", "beta2", "Slope", "Intercept_1", "Intercept_2")
for (i in param) {
  d_elbow <- dd_boot_long_fine |>
    filter(Dilution == "50pg & 25pg") |> 
    group_by(Parameter, Samplesize) |> 
    summarise("STD" = sd(Estimate)) |> 
    ungroup() |> filter(Parameter == i) |>
    select(Samplesize, STD)
  
  dd_elbow <- (d_elbow$Samplesize[-1] + d_elbow$Samplesize[-nrow(d_elbow)])/2
  dd_elbow <- cbind(dd_elbow, d_elbow[[2]][-nrow(d_elbow)] - d_elbow[[2]][-1])
  
  print(i)
  plot(dd_elbow[,1], dd_elbow[,2])
  abline(h=(max(dd_elbow[,2])+min(dd_elbow[,2]))/2)
}
```


Note: The data contains around $3/4$ homozygous genotypes and $1/4$ heterozygous genotypes, but the majority of the wrong calls are found among the heterozygous genotypes. A heterozygous genotype is generally more prone to become a wrong call, since one of its alleles can drop out leaving only the signal from the other allele (while an allele drop out for a homozygous genotype still leaves an intact signal from the other allele which is of the same kind).
```{r}
dils2 <- unique(dd_GT_both$Dilution)[order(unique(dd_GT_both$Dilution), decreasing = T)]
model_dil <- lapply(dils2, function(d) {
  df <- dd_GT_both |> filter(Dilution %in% d)
  m <- fit_symmetry_model2(df, f = function(x) sqrt(x), intercept = TRUE, method = "Nelder-Mead", control_list = list(maxit = 500))
  df <- predict_prob_threshold(df, m)
  return(df)
}) |> bind_rows()
# names(model_dil) <- paste0(dils2, "pg")

model_dil |> group_by(Dilution) |> 
  summarise("%Het" = round(sum(Genotype_true_AA == "A1A2")/n()*100, digits = 1),
            "%Hom" = round(sum(Genotype_true_AA != "A1A2")/n()*100, digits = 1),
            "%WC" = round(sum(Genotype_true != Genotype_pred)/n()*100, digits = 2),
            "#WC|Het" = sum((Genotype_true != Genotype_pred) & (Genotype_true_AA == "A1A2")),
            "#WC|Hom" = sum((Genotype_true != Genotype_pred) & (Genotype_true_AA != "A1A2")),
            "%WC among Het" = round(sum((Genotype_true != Genotype_pred) & (Genotype_true_AA == "A1A2"))*100/sum(Genotype_true_AA == "A1A2"), digits = 2),
            "%WC among Hom" = round(sum((Genotype_true != Genotype_pred) & (Genotype_true_AA != "A1A2"))*100/sum(Genotype_true_AA != "A1A2"), digits = 2)) |> 
  arrange(desc(Dilution))
```