\begin{table*}
\centering
\caption{The SMLR model's genotyping improvements for all examined DNA quantities.} 
\label{tab:NC_WC}
\vspace{1mm}
\begin{adjustbox}{max width=\textwidth}
\begin{threeparttable}
\begin{tabular}{llrrrrrrrrrr}
  \toprule[2pt]
\null & \null & \multicolumn{6}{c}{First series:} & \multicolumn{4}{c}{Second series:} \\
\null & \null & \multicolumn{6}{c}{5 individuals, 4 examinations} & \multicolumn{4}{c}{18 individuals, 1 examination} \\
\cmidrule(lr){3-8} \cmidrule(lr){9-12}
\multicolumn{2}{l}{DNA quantity (\si{\pg})} & 1,000 & 500 & 250 & 125 & 62.5 & 31.25 & 50 & 25 & 12.5 & 6.25 \\ 
\multicolumn{2}{l}{Number of SNPs} & 3,240 & 3,240 & 3,240 & 3,238 & 2,585 & 1,931 & 2,831 & 2,821 & 2,575 & 2,347 \\ 
\midrule[2pt]
\multirow{2}{*}{HSG\tnote{$*$}} & No-calls & 1 & 3 & 28 & 48 & 83 & 156 & 62 & 142 & 213 & 276 \\ 
   & Wrong calls & 0 & 0 & 1 & 1 & 2 & 9 & 8 & 59 & 184 & 272 \\ 
\midrule[2pt]
\multirow{5}{*}{SMLR\tnote{$\dagger$}}  & $q$ aligning WCs & 0 & 0 & 0 & 0.5473 & 0.5473 & 0.7394 & 0.8115 & 0.8449 & 0.8302 & 0.8541 \\ 
\cmidrule(lr){2-12}
   & Call rate (\%) & 100 & 100 & 100 & 99.8 & 99.7 & 96.0 & 98.4 & 95.7 & 95.1 & 92.2 \\ 
   & $\text{No-calls}^{\ddagger}$ & 0 & 0 & 0 & 5 & 7 & 78 & 46 & 121 & 125 & 184 \\ 
   & NC-reduction (\%)\tnote{$\S$} & 100 & 100 & 100 & 89.6 & 91.6 & 50.0 & 25.8 & 14.8 & 41.3 & 33.3 \\ 
\midrule[2pt]
\multirow{5}{*}{SMLR\tnote{$\ddagger$}} & $q$ aligning NCs & 0.8890 & 0.8808 & 0.8588 & 0.8592 & 0.8514 & 0.8414 & 0.8488 & 0.8592 & 0.8725 & 0.8841 \\ 
\cmidrule(lr){2-12}
   & Accuracy (\%) & 100 & 100 & 100 & 100 & 100 & 99.775 & 99.747 & 97.878 & 92.760 & 88.048 \\ 
   & Wrong calls & 0 & 0 & 0 & 0 & 0 & 4 & 7 & 57 & 171 & 248 \\ 
   & WC-reduction (\%)\tnote{$\S$} & 0 & 0 & 100 & 100 & 100 & 55.6 & 12.5 & 3.4 & 7.1 & 8.8 \\ 
   \bottomrule[2pt]
\end{tabular}
\begin{tablenotes}
\item[$*$] Number of no-calls (NCs) and wrong calls (WCs) made by the HID SNP Genotyper Plugin (HSG).
\item[$\dagger$] The SMLR model's performance when $q$ is set to align WCs according to \eqref{eq:metric_CR}.
\item[$\ddagger$] The SMLR model's performance when $q$ is set to align NCs according to \eqref{eq:metric_AC}.
\item[$\S$] Results are conservative and based on the SMLR model with an intercept fitted to square-root transformed allele signals from examinations of \SI{25}{\pg} and \SI{50}{\pg} DNA.
\end{tablenotes}
\end{threeparttable}
\end{adjustbox}
\end{table*}
